encoder.go

```diff
@@ -0,0 +1,152 @@
+package schema
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+type encoderFunc func(reflect.Value) string
+
+// Encoder encodes values from a struct into url.Values.
+type Encoder struct {
+	cache  *cache
+	regenc map[reflect.Type]encoderFunc
+}
+
+// NewEncoder returns a new Encoder with defaults.
+func NewEncoder() *Encoder {
+	return &Encoder{cache: newCache(), regenc: make(map[reflect.Type]encoderFunc)}
+}
+
+// Encode encodes a struct into map[string][]string.
+//
+// Intended for use with url.Values.
+func (e *Encoder) Encode(src interface{}, dst map[string][]string) error {
```

Comment:
nitpicking is no problem. I'm a little unclear about your ask here, though. `url.Values` is an alias of `map[string][]string` and the latter agrees with the Decoder's signature. Would you rather use `url.Values` for all of the public APIs here? I agree with not returning a map allocated internally.